### DISCUSSION TITLE
When does app.storage.tab delete a user’s unique storage dicts?

### DISCUSSION THREAD

**Discussion by JS-Aibel** [zauberzeug/nicegui#3037] [Q&A]
### Question

**Question**
When does **app.storage.tab** delete the storage dicts that are unique to each tab session?
When the user closes the tab, will the stored server-side in memory data that the user has allocated be deleted?

From the app.storage.tab pull request:
> keeps a state dictionary as long as the browser tab lives.

Does this mean that the data itself is deleted from the server, or only that the pointer to the data is deleted from the user’s perspective?

**Reference**
In the documentation [Storage | NiceGUI](https://nicegui.io/documentation/storage#storage) it says:
- “Data will be lost when restarting the server”
- “in app.storage.tab which can be persisted on the server even for days”

The [pull request](https://github.com/zauberzeug/nicegui/pull/2837) says:
- “a volatile storage which keeps a state dictionary as long as the browser tab lives.”

**Comment by rodja** [comment]
The tab storage of an individual tab is removed after 30 days of inactivity (or when the the server restarts):

https://github.com/zauberzeug/nicegui/blob/cf61dcb2191d55fbf38fb8d38fc7867a4801c07c/nicegui/storage.py#L188-L194

The max age can be configured to be only one day with `app.storage.max_tab_storage_age = timedelta(days=1).total_seconds()`.

**Reply by JS-Aibel** [reply]
Thank you for the answer, @rodja.
Why is not the data deleted when the user closes the tab also? Allowing the server to keep data in memory that is no longer in use sounds like a memory leak.

**Reply by rodja** [reply]
I do not think there is a way to detect the closing of a tab.
